### PR TITLE
[E] Whisper

### LINKED ISSUE #7: Write testcases for everything under internal

### LINKED ISSUE #11: Implement transcribing using the whisper api
I tested out whisper and wow, it's much faster and much more accurate the google's cloud speech api. Implementing this will require a rework and reorganisation of the speech module so as to avoid spaghettification.

### PR DESCRIPTION
**Pull request by mmatongo**
closes #11 
partially resolves #7 
begins work on #9

### CODE REVIEW

File: internal/transcribe/gcs_utils.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			},
+		},
+	}
+}
+
+func extractTranscript(resp *speechpb.LongRunningRecognizeResponse) string {
+	var transcript string
+	for _, result := range resp.Results {
+		for _, alt := range result.Alternatives {
+			transcript += alt.Transcript
```
Comment (ccoVeille) [inline internal/transcribe/gcs_utils.go:94]: No separator?

I mean like a space or \n

File: internal/transcribe/google_transcriber.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package transcribe
+
+import (
+	speech "cloud.google.com/go/speech/apiv1"
+	"cloud.google.com/go/storage"
+	"context"
+	"fmt"
+	"path/filepath"
+)
```
Comment (ccoVeille) [inline internal/transcribe/google_transcriber.go]:
```suggestion
import (
	"context"
	"fmt"
	"path/filepath"

	speech "cloud.google.com/go/speech/apiv1"
	"cloud.google.com/go/storage"
)
```

File: internal/transcribe/gcs_utils.go (hunk 1 above)
Comment (mmatongo) [reply internal/transcribe/gcs_utils.go:94]: No, the text comes out already formatted albeit poorly in some cases

File: internal/transcribe/google_transcriber.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	client, err := createSpeechClient(ctx, opts)
+	if err != nil {
+		return "", err
+	}
+	defer func(client *speech.Client) {
+		err := client.Close()
+		if err != nil {
+			fmt.Printf("failed to close transcribe client: %v\n", err)
+		}
+	}(client)
```
Comment (ccoVeille) [inline internal/transcribe/google_transcriber.go:24]: The fmt.Printf is a bit strange to me

The defer is called before returning the value

So you could do something like this


```suggestion
	defer func(client *speech.Client) {
		errCloser := client.Close()
		if errCloser != nil {
			err = errors.Join(err, fmt.Errorff("failed to close transcribe client: %w", errCloser)
		}
	}(client)
```

Comment (ccoVeille) [reply internal/transcribe/google_transcriber.go:24]: The code might need more rewriting of course.

Another solution is to extract the code calling everything needed before the defer in a lamda function, then call the code currently in the defer after the lamda function

Comment (mmatongo) [reply internal/transcribe/google_transcriber.go:24]: I'm a little confused, doesn't the actual execution of the deferred function happen just before the containing function returns?

Your code is great by the way, I didn't know about errors.Join

Comment (ccoVeille) [reply internal/transcribe/google_transcriber.go:24]: This article explains a bit

https://trstringer.com/golang-deferred-function-error-handling/

Here is an article about using defer plus errors.Join

https://wstrm.dev/posts/errors-join-heart-defer/

**Pull request merged**